src/raw/owned_bson.rs

```diff
@@ -0,0 +1,504 @@
+    Int32(i32),
+    /// 64-bit signed integer
+    Int64(i64),
+    /// Timestamp
+    Timestamp(Timestamp),
+    /// Binary data
+    Binary(Binary),
+    /// [ObjectId](http://dochub.mongodb.org/core/objectids)
+    ObjectId(oid::ObjectId),
+    /// UTC datetime
+    DateTime(crate::DateTime),
+    /// Symbol (Deprecated)
+    Symbol(String),
+    /// [128-bit decimal floating point](https://github.com/mongodb/specifications/blob/master/source/bson-decimal128/decimal128.rst)
+    Decimal128(Decimal128),
+    /// Undefined value (Deprecated)
+    Undefined,
+    /// Max key
+    MaxKey,
+    /// Min key
+    MinKey,
+    /// DBPointer (Deprecated)
+    DbPointer(DbPointer),
+}
+
+impl OwnedRawBson {
+    /// Get the [`ElementType`] of this value.
+    pub fn element_type(&self) -> ElementType {
+        match *self {
+            OwnedRawBson::Double(..) => ElementType::Double,
+            OwnedRawBson::String(..) => ElementType::String,
+            OwnedRawBson::Array(..) => ElementType::Array,
+            OwnedRawBson::Document(..) => ElementType::EmbeddedDocument,
+            OwnedRawBson::Boolean(..) => ElementType::Boolean,
+            OwnedRawBson::Null => ElementType::Null,
+            OwnedRawBson::RegularExpression(..) => ElementType::RegularExpression,
+            OwnedRawBson::JavaScriptCode(..) => ElementType::JavaScriptCode,
+            OwnedRawBson::JavaScriptCodeWithScope(..) => ElementType::JavaScriptCodeWithScope,
+            OwnedRawBson::Int32(..) => ElementType::Int32,
+            OwnedRawBson::Int64(..) => ElementType::Int64,
+            OwnedRawBson::Timestamp(..) => ElementType::Timestamp,
+            OwnedRawBson::Binary(..) => ElementType::Binary,
+            OwnedRawBson::ObjectId(..) => ElementType::ObjectId,
+            OwnedRawBson::DateTime(..) => ElementType::DateTime,
+            OwnedRawBson::Symbol(..) => ElementType::Symbol,
+            OwnedRawBson::Decimal128(..) => ElementType::Decimal128,
+            OwnedRawBson::Undefined => ElementType::Undefined,
+            OwnedRawBson::MaxKey => ElementType::MaxKey,
+            OwnedRawBson::MinKey => ElementType::MinKey,
+            OwnedRawBson::DbPointer(..) => ElementType::DbPointer,
+        }
+    }
+
+    /// Gets the wrapped `f64` value or returns `None` if the value isn't a BSON
+    /// double.
+    pub fn as_f64(&self) -> Option<f64> {
+        match self {
+            OwnedRawBson::Double(d) => Some(*d),
+            _ => None,
+        }
+    }
+
+    /// Gets a reference to the `String` that's wrapped or returns `None` if the wrapped value isn't
+    /// a BSON String.
+    pub fn as_str(&self) -> Option<&'_ str> {
+        match self {
+            OwnedRawBson::String(s) => Some(s),
+            _ => None,
+        }
+    }
+
+    /// Gets a reference to the [`RawArrayBuf`] that's wrapped or returns `None` if the wrapped
+    /// value isn't a BSON array.
+    pub fn as_array(&self) -> Option<&'_ RawArray> {
```

Comment:
Since there are now mutating methods for those, I think it'd be good to have `as_array_mut -> Option<&mut RawArrayBuf>` and similar for `as_document_mut`.

Reply:
good idea, added